pkg/serviceaccounttoken/secret.go

```diff
@@ -204,29 +233,89 @@ func isSecretForServiceAccount(secret *corev1.Secret, sa *corev1.ServiceAccount)
-	mutex := getLock(lockKey)
-	mutex.Lock()
-	defer func(key string) {
-		mutex.Unlock()
-		lockMap.Delete(lockKey)
-	}(lockKey)
-
-	// We could have been waiting for the Mutex to unlock in a parallel run of
-	// createServiceAccountSecret - check again for the secret existing.
-	secret, err := ServiceAccountSecret(ctx, sa, secretLister, secretClient)
+func createServiceAccountSecret(ctx context.Context, sa *corev1.ServiceAccount, secretLister secretLister, secretClient clientv1.SecretInterface) (*corev1.Secret, error) {
+	sc := SecretTemplate(sa)
+	secret, err := secretClient.Create(ctx, sc, metav1.CreateOptions{})
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error ensuring secret for service account %s: %w", logKeyFromObject(sa), err)
 	}
-	if secret != nil {
-		return secret, nil
+
+	return secret, nil
+}
+
+// returns true if the secret has changed.
+func annotateSAWithSecret(ctx context.Context, sa *corev1.ServiceAccount, secret *corev1.Secret, saClient clientv1.ServiceAccountInterface, secretClient clientv1.SecretInterface) (*corev1.ServiceAccount, bool, error) {
+	if sa.Annotations == nil {
+		sa.Annotations = map[string]string{}
 	}
+	secretAnnotation := secret.Namespace + "/" + secret.Name
 
-	sc := SecretTemplate(sa)
-	secret, err = secretClient.Create(ctx, sc, metav1.CreateOptions{})
+	// If the SA is already annotated with the secret already
+	if ann := sa.Annotations[ServiceAccountSecretRefAnnotation]; ann == secretAnnotation {
+		return sa, false, nil
+	}
+
+	sa.Annotations[ServiceAccountSecretRefAnnotation] = secretAnnotation
+
+	updated, err := saClient.Update(ctx, sa, metav1.UpdateOptions{})
+	if err == nil {
+		return updated, false, nil
+	}
+	if !apierrors.IsConflict(err) {
+		logrus.Debugf("Failed to update ServiceAccount for %s: %s", logKeyFromObject(sa), err)
+		return nil, false, err
+	}
+
+	// Rollback the optimistically created secret
+	logrus.Debugf("Rolling back ServiceAccount secret for %s", logKeyFromObject(secret))
+	if err := secretClient.Delete(ctx, secret.Name, metav1.DeleteOptions{}); err != nil {
+		return nil, false, fmt.Errorf("deleting optimistically locked secret for %s - %s: %w", logKeyFromObject(secret), logKeyFromObject(sa), err)
+	}
+	// Load the version that triggered the issue
+	logrus.Debugf("Reloading ServiceAccount %s", logKeyFromObject(sa))
+	updated, err = saClient.Get(ctx, sa.Name, metav1.GetOptions{})
 	if err != nil {
-		return nil, fmt.Errorf("error ensuring secret for service account [%s:%s]: %w", sa.Namespace, sa.Name, err)
+		return nil, false, fmt.Errorf("getting updated service account %s: %w", logKeyFromObject(sa), err)
 	}
 
-	return secret, nil
+	return updated, true, nil
+
+}
+
+// This should be replaced with a types.NamespacedName which stringifies to
+// <namespace>/<name> but that would change the log output format.
+func logKeyFromObject(obj namespacedObject) namespacedName {
```

Comment:
Nit: I'm not sure that we need the extra overhead of this type for a few reasons:
- If we really want to have a separate function for the key due to re-use, then we should just return the output of the string function, rather than include the overhead of the string representation.
- I don't think that the exact interface of the key in the log statement is an API-level construct that we need backward compat for.

Reply:
I added it to avoid changing the output format, `types.NamespacedName` would be a better option.

This is useful mostly to keep the same ordering in output convention i.e. it's always `namespace/name`.

Reply:
I've stripped it in favour of a function that returns `namespace:name`.